### ISSUE TITLE
[Suggestion] Botania, Integration in Twilight Forest, Integration Addons

### ISSUE THREAD

**Issue by Rosenlied** [GTNewHorizons/GT-New-Horizons-Modpack#6446]
I have looked through previous proposals of Botania, but I have yet to see the use of Botania as magic gated entirely behind Twilight Forest progression as a possibility. Proposal: Botania being introduced and locked behind the Twilight Forest's castle boss. 

**Notes on unique features that Botania (base) adds:**

-Alternatives to pure-tech powered tools: Manasteel tools provide "early" access to powered tools albeit at a lower tier, Elementium tools are unique in functionality/auto voiding ability, Terrasteel tools provide magical AoE options and an alternative Lumber Axe. 

-Additional options for mob farming: Bellethorn is a powered iron spike alternative, while Hyacidus applies poison effect to mobs to assist with diamond spikes or similar. The Life Aggregator provides an alternative means of moving spawners besides the Diamond Dolly (a 'cheaper' but further gated, one time use recipe.)

-Alternatives to bees for clay generation: Clayconia flower turns sand into clay balls directly. May require gating to avoid trivializing lithium production.

-Alternative enchanting method not provided by Blood Magic/Thaumcraft, uses enchanted books but does not consume them. Can be gated as necessary (e.g. changing multiblock to use Obsidian Tiles and a Runic Matrix instead of obsidian/lapis blocks)

-Assembly Halo - Forestry Worktable alternative, can be carried.

-Alternatives for automated farming of vanilla-style (Harvestcraft, Thaumic Bases) crops and trees, block breakers.

-A magical jetpack that provides creative-style flight, the Flugel Tiara. Has a short limit on duration.

-Red Stringed items: additional "remote block faces" for interacting with machines.

-Corporea System: An alternative to Project Red for autocrafting, apparently less laggy in certain applications though does not support liquids. 

-Orechid and Loonium flowers, used to generate ores and dungeon loot. See below for more notes (if added these should be gated heavily but should still be usable).

**Notes on existing Botania/Thaumcraft integration:**

-Manasteel/Elementium caps and Livingwood/Dreamwood rods: thematic alternatives to Blood Magic-based wand components.

-Flux scrubbing flower, research aspect generating flower, and ethereal essence generating flower

**Notes on existing Botania/Forestry integration:**

-Botania based grafters/scoops

-Botania mana based apiary stuff

-Flowers that turn ignoble bees pristine, directly generates bees, and consume surplus drones

**Notes on planned features for Botania integration with GregTech and other mods**

I'm currently working on developing a Botania addon, Garden of Avalon, which expands on Botania in a way that fits with GregTech. Some planned features:

-Flower that provides items to nearby machines (+flowers): intended to reduce lag from items dropped in world that Botania likes to use. 

-"Flower Trellis" system, allowing for advanced multiblock Botania functional flowers as well as singleblock flowers to be gated further using technology.

-Additional components to ore processing system: Orechid tweaks to require alternative ore processing methods, likely using a highly gated mining method.

-Alternative generating flowers/multiblock flower bushes to fit a more GregTech style of power generation

Keep in mind that Botania would naturally be gated behind defeating the castle boss of the Twilight Forest, so any gating that would apply to that boss would also apply to Botania.

**Comment by Cleista** [comment]
Here are my partial comments from the perspective of GTNH pack: 

> Proposal: Botania being introduced and locked behind the Twilight Forest's castle boss.

It's not really a meaningful gate (progression-wise) because once you're in TF you can complete all of it without teching up further. From the player perspective this is just tedious, especially since a lot of Botania's utility is targeted towards early game. It's why GTNH wand progression requires both TF parts **and** GT materials - it's a gradual TF progression requirement and the actual gating is by your GT tier. 

> Alternatives to pure-tech powered tools: Manasteel tools provide "early" access to powered tools albeit at a lower tier, Elementium tools are unique in functionality/auto voiding ability, Terrasteel tools provide magical AoE options and an alternative Lumber Axe.

In GTNH most people just use TiCon picks and hammers and largely avoid electric tools, I'm not sure if they'll make Manasteel tools (and if added should make sure Botania tool mining level doesn't let you skip ore hardness tiers). 

The Terrasteel tools (if ranked up sufficiently) are functionally superior to TiCon hammers (and arguably even awakened ichorium) so they should be Minetweakered to tier up using GT materials rather than mana. I like the way Infitech did it: infusion recipes with various different GT materials to increase tier.

> Alternatives to bees for clay generation: Clayconia flower turns sand into clay balls directly. May require gating to avoid trivializing lithium production.

People can get tons of clay from mesas right now, it's fully renewable from Bio Chaff -> Dirt -> Mud -> Clay balls, and Blood Magic lets you make it from sand. So clay generation probably doesn't need much of a nerf. 

> Alternative enchanting method not provided by Blood Magic/Thaumcraft, uses enchanted books but does not consume them. Can be gated as necessary (e.g. changing multiblock to use Obsidian Tiles and a Runic Matrix instead of obsidian/lapis blocks)

Enchanted book duping via bibliocraft is already fairly accessible, so this isn't particularly OP with respect to what's already in the pack. Changing the multi for flavor purposes probably isn't a bad idea if it's easy, though idk whether or not Runic Matrix is a TE that would tick and cause lag (if so probably use something else). 

> Assembly Halo - Forestry Worktable alternative, can be carried.

It's great QoL and not too OP, would be nice to have access to it in MV. 

> Alternatives for automated farming of vanilla-style (Harvestcraft, Thaumic Bases) crops and trees, block breakers.

Assuming it's not particularly laggy could be a nice addition. Better than golems. 

> Red Stringed items: additional "remote block faces" for interacting with machines.

Tier it similarly to RemoteIO/Transvector Interface and should be fine. 

> Orechid and Loonium flowers

They need a running cost that isn't mana, if possible to implement. 

* Orechid: 250L UUM (maybe a different amount per dimension and have it generate only ores that generate in that dimension)
* Loonium: 1L Molten Infinity

> I'm currently working on developing a Botania addon, Garden of Avalon...

Your GoA addon ideas sound really nice. Would love to see it in action. 

-----

Regarding the mod in general: 

Mana generation needs to be completely overhauled. Botania mana gen is balanced around vanilla so mana is basically produced at almost no cost in any mod pack. I suggest balancing as follows: 

* Remove all mana generation flowers. 
* Change mana cost so it has a meaningful resource consumption. Enable a TC Alchemy recipe for black lotus using a relatively non-renewable catalyst (shard clusters?). Infusion for Blacker Lotus at Ichorium tier maybe. 


The following methods don't work too well IMO: 

1. Massively increasing mana costs for everything/decreasing mana generation rate
    * Tried this with another pack, people just expanded their mana generation causing TPS lag and doesn't fundamentally address the fact that mana is fully renewable and costs practically nothing. 
2. Making EU charge Mana Tablet
    * Unless the ratio was really high (i.e. 2048 EU = 1 mana) it would not solve the problem of mana being practically free. 
3. Gating functional flowers significantly
    * The recipes need balancing, but should be appropriate for their function and not excessively hard. Without the mana change, it doesn't address the problem of negligible running cost and will just make people not do Botania.

**Comment by Rosenlied** [comment]
> Remove all mana generation flowers

This feels somewhat excessive, to do with all mana generating flowers. Stuff like Endoflames are trivially automated, so it'd be reasonable to disable those.
Something such as the Entropynnium is less trivial: this would be an additional scaling of an explosives production plant in addition to what is needed for implosion compressors.

I'm also planning on Garden of Avalon flowers being either a challenge to handle or at least mirror some Gregtech things. Still no direct electricity-to-mana conversion, however.
An example of the former would be a flower that generates mana according to a cellular automaton that is not the standard Game of Life, as a 'reactor' type, while the latter would include a flower that is essentially a "magical large turbine" that converts steam/liquids into mana as well as a Thaumcraft-gated flower that generates mana from centivis.

Flowers that definitely deserve to be cut completely including from NEI would be stuff such as all passive flowers, the Endoflame (use steam flower + boilers), Thermalily (steam flower + lava boiler), Gourmaryllis, Rosa Arcana, and Kekimurus. Kekimurus is questionable but can be adjusted to instead only accept Red Velvet Cake as the Aer Infused Fire creates vanilla cakes.

The Tainthistle is a generating flower but by its nature is difficult to use as primary generation due to eating flux, and is primarily a flux scrubber. Gating should be at Apocrypha tab, behind Fluxium. Beegonia is questionable, but likely should be disabled as a simple Ender IO conduit and 4x fertility cultivated bees would allow for trivial drone production to feed it.

Narslimmus can only consume slimes generated via slime chunk, which is itself somewhat of a prospecting task. Spectrolus may be interesting to adjust to emit a signal depending on the color it requests and randomize and require a redstone system to supply it with the proper color, but this may be a lag source so may need removal. Dandelifeon is a nontrivial automation task and 'reactor-like', and is naturally gated behind Gaia Spirits and is thus naturally heavily gated. Entropynnium, again, is a flower that uses TNT which is a nontrivial automation task.

Making mana production completely unrenewable would be an issue and strongly discourage use of Botania. The intention here would be to, rather than nerfing it into unusability, make mana generation be more comparable with existing power generation options. In essence, making it as "practically free" as EU (read: not really) while not convertable.

**Comment by Cleista** [comment]
Entropinnyum has [exploits with vanilla TNT duping](https://ftb.gamepedia.com/Entropinnyum#Tips) although the other flower generators could be fine going by what you're saying here. (Spectrolus probably won't be that laggy actually, would be interesting to see how it gets automated).

Note that you can make obscene amounts of centivis in GTNH pretty trivially, so mana generation from it may not be balanced unless the flower is fairly late game.

As long as mana is a resource whose usage people actually need to care about (like how Essentia feels early game) it would go a long way to improve the ability to balance Botania relative to the rest of the pack. The other thing to investigate would be a way to set non-mana running costs for functional flowers (specifically a fluid cost) which will solve a lot of balancing concerns.

**Comment by Rosenlied** [comment]
> Orechid and Loonium issue

Directly giving these flowers a running cost directly besides mana looks like it'd be a difficult option. An equivalent option would be making it so the Orechid and Loonium produce resources that must be processed further, ideally using Botania-derived stuff.

Simply requiring UU matter or Molten Infinity doesn't add additional depth to these flowers besides just "more numbers", however.

For the Orechid:

-Orechid (dimension variant) - generates a drop table that lines up with one of the dimension's large ore veins (weight 30 for primary/secondary, 25 for between, 10 for sporadic? Weights can be adjusted) when made, and enriches gravel into Alluvial Gravel corresponding to the ore. Must be made and used in the appropriate dimension.
Alluvial Gravel itself has to be processed into Enriched Gravel using a Botania+Witchery based sluice structure that uses Flowing Spirit to sluice out the minerals.
Enriched Gravel would be processed by way of Alchemical Flotation Cell, a Thaumcraft multiblock that consumes large amounts of Aer/Aqua/Terra centivis (which must be provided by a single node: a requirement of n centivis requires n^2 aspects in the pre-energized node) to produce Alchemical Froth.
Alchemical Froth would be then processed in the Alchemical Refinery, using Blood Magic Life Essence in a distillation tower like structure to separate into individual ore dusts. 

(Here, all magic mods and progression within all of them would be required to be used, and require additional infrastructure to even be able to use the ores produced here, while adding more complexity to the process of Orechid ore processing than simply adding UU matter to the equation.)

As for the Loonium, making it take mana to generate an item that has to be refined further and spawns a boss (dimension restricted to a space dimension) that generates a dungeon loot chest when killed would be a potentially viable means of balancing it, especially depending on what is used to refine the item further. 

Essentially: balancing by rather than adding more numbers, adding more processing chain (which require high tier materials for the parts).

**Comment by Rosenlied** [comment]
>Entropinnyum has exploits with vanilla TNT duping 

Vanilla TNT duping is only possible using slime block frames, which are not possible in 1.7.10. As for fluid costs for flowers, I'm not sure which flowers would need fluid costs that don't generate items (where additional refinement costs could be handled with the processing of the items it makes) and it seems like it would be easier to handle that way.

**Comment by yukieiji** [comment]
My thoughts are as follows.

> Corporea System: An alternative to Project Red for autocrafting, apparently less laggy in certain applications though does not support liquids.

This can easily make item transport and redstone wireless, so need to gate to a higher tier (EV or HV?)

> Alternative enchanting method not provided by Blood Magic/Thaumcraft, uses enchanted books but does not consume them. Can be gated as necessary (e.g. changing multiblock to use Obsidian Tiles and a Runic Matrix instead of obsidian/lapis blocks)

When used in conjunction with an overenchanter, nerfs are necessary as they allow for the mass production of enchanted level weapons/armor not normally obtainable.

"Ring of Loki" and "Ring of Odin" , "Eye of the Frugel" are OP.

I think that Botania should be gated by TC's research and GregTech. so this would prevent mass production of low cost mana generation flowers and the OP items from getting to them early.

**Comment by Rosenlied** [comment]
> I think that Botania should be gated by TC's research and GregTech. so this would prevent mass production of low cost mana generation flowers and the OP items from getting to them early.

The low cost mana generation flowers would be disabled entirely, and mana generation would be pushed to either GregTech generator like Garden of Avalon flowers or more difficult to automate mana generation flowers. OP flowers definitely would be gated through both GregTech to limit availability and in the case of the Orechid, processing chains requiring late game Thaumcraft/Witchery/Blood Magic. It can be seen as a potential variant void miner that takes additional automation steps, and as such should be balanced as such (e.g. the Garden of Avalon processing chain I'm working on.)

The Gaia Spirit Ingot naturally gates Ring of Loki/Ring of Odin (and by extension Ring of the Aesir) and Eye of the Flugel, though Gaia Guardian Relics can be gated further by requiring something like assembly line to repair the dropped, broken relics.

**Comment by Dream-Master** [comment]
I looking now videos to know how Botnia works. We not need to disable low flowers but we can change tha values. I think lets go into the code and change the recipes as we need it and only add less things to minetweaker. We can change and adjust allrecipes better from inside the mod. But the more important questions is should weadd this mod in the first place ?

**Comment by Cleista** [comment]
I really like the processing chain balancing idea. It fits the theme of the pack and solves the problem of needing a vector for resource consumption balancing. 

> Vanilla TNT duping is only possible using slime block frames, which are not possible in 1.7.10. 

That makes it a bit more secure. Since vanilla ripped off the frames from mods anyway I think duping TNT is probably possible still, but since there's no easily accessible proof of concept the likelihood of abuse is low. 

@Dream-Master Like I mentioned previously, changing values will just make people spam more flowers, killing tps. Any mana cost nerf without fixing the generation just results in chunks full of Endoflames or autocrafted Hydroangeas. Doesn't solve the problem and just makes server performance worse. 

The problem with mana in Botania is that it's required for everything but the consumption is negligible and there's no real generation tiering. You just make a mana generation setup once and it's enough for everything forever. All the transmutation/alchemy/crafting recipes that use mana? Might as well spawn in an everlasting guilty pool for every player, because a unit of mana costs nothing to generate. 

1 million mana should be at least as hard to get as 1 billion EU. Make the generation accessible but expensive to sustain, and we can start expressing costs in terms of **the mana generation cost** e.g. "one mana diamond is approximately 20 TNT" which makes it much easier to balance recipes that cost mana as well as balancing mana generation flowers relative to one another. ("Should 20 TNT and 10 Red Velvet Cake be equivalent for mana generation?")

Note that Botania has worldgen, so for compatibility with old servers, turning off worldgen and giving a TC way to craft floral fertilizer is the most sensible way to gate startingbotania.

**Comment by Dream-Master** [comment]
http://jenkins.usrv.eu:8080/view/3rd%20Party%20Mods/job/Botania/2/

**Comment by Dream-Master** [comment]
so i fork it make it buildable and put it to my jenkins.now we can work on it. If it fits some day we will add it.

**Comment by Dream-Master** [comment]
https://github.com/GTNewHorizons/Botania

**Comment by Regian24** [comment]
well maybe late game flower that generates mana from like naquadah?

**Comment by Dream-Master** [comment]
We need to rework the whole mod and change all recipes not just a singe flower

**Comment by Regian24** [comment]
Yeah

**Comment by TataTawa** [comment]
I use animated-texture-fix mod which is a tiny mod and makes tick time cut down by ~30% if your GPU supports ARB_COPY_IMAGE (or Opengl 4.x)
Unfortunately if use Botania, animated-texture-fix, JourneyMap together there is conflict. Disable animated-texture-fix function in config will be a work-around.

to completely fix this conflict, just makes code here
vazkii.botania.client.render.block.InterpolatedIcon (which extends vanilla TextureAtlasSprite)
extends
https://github.com/Kobata/animated-texture-fix/blob/master/src/main/java/animfix/FastTextureAtlasSprite.java

that mod also needed to add in pack

**Comment by Irgendwer01** [comment]
My Idea for integrating Botania is removing most generating flowers and replace them through equivalent ones that generate Mana with GT Fuels like Oil and Diesel and make Botania materials like Manasteel obtainable via EBF with processing steps like this:
Steel Dust -> Mana Pool -> Manasteel Dust -> EBF -> Manasteel Ingot
Hope im not too late on this discussion lol

**Comment by Chaoschaot232** [comment]
> My Idea for integrating Botania is removing most generating flowers and replace them through equivalent ones that generate Mana with GT Fuels like Oil and Diesel and make Botania materials like Manasteel obtainable via EBF with processing steps like this: Steel Dust -> Mana Pool -> Manasteel Dust -> EBF -> Manasteel Ingot Hope im not too late on this discussion lol

To be honest but the EBF has nothing todo with Magic so making magic stuff in the EBF is worse which I would question as the player playing this pack.

Generally spoken:

It is fine to gate stuff behind a Tier. It is fine to use GT Mats for there Recipes. But only because GT itself is in the pack is not a freeway ticket to make such deep changes to destroy the entiere flavor of Botania. Such a change would for me personaly end up in reverting back changes which is possible.

Sometimes to many ppl. are thinking in a to much "gregifying" way because of "Oh no, this stuff is so OP, we must destroy the mods flavor compleatly."

1. Use the TF for Flower Generation
2. Use the Portal to Alfheim and trade suiteable GT Materials in it to get Manasteel, Alfheim Glass and so on. Removing the duplication recipies would be fine as they are not needed.
3. Tweak the amount of Mana needed for some recipies or maybe tweak the amount of generated Mana but do not touch Botanias flare in terms of how and with what Mana is generated.
4. Terrasteal tools should be done with Thaumcraft, requiring GT Materials, TF Materials and Botania Materials. In addition they should also be marked as Forbidden (Taboo) with a complicated Minigame and so on (will show such a recipe soon enough).
5. To level up the OP Terra Shatterer, which uses Mana from a held Mana Tablet in your inventory to work and eats up your hunger like nothing, let it eat up a few more times of Mana and if possible, stop at a cerain amount and tell the player that he has to infuse it with better materials to level up that tool. This would not only make it mana hungry enough, it would also require much more expensive materials from GT to gate it's mane functionality behinded some tiers. Maybe smth. like Naquadah to get it a Rank A or B once (Ranks are going from D to SS) and well, it's Mininglevel is set to 06-Ardite so it can't handle stuff (even my current hammer can only handle 06-Ardite ores for now).
6. Split down the Mana Pylon and Gaya Pylon into two recipes as they are not only a crystal, like the Natura Pylon is. They have a ring in addition which could be created to make it later craftable as Infusion. Make than an upgrade infusion where the Mana Pylon Crystal and the Mana Pylon Crystal Ring are infused in the Natura Pylon with GT Materials and bunch of other suitable stuff (will show smth. soon to) and do the same for the Gaia Pylon (but make this one Forbidden (Taboo).
7. Change the Runic Altar recipies with higher Mana Costs and not only adding one single pice of Stainless Steel ingot. Maybe it is by now only a place holder but this is definitly the wrong way as it would looks like a late april fools joke, tbh.

Well, well - there is many more to say but I could write pages after pages the hole day so instead of this I am going to make and show my changes and ideas to suite both: The GTNH Flavor and the Botania Flavor so that we can discuss over them, thanks :) - This will defninitly take a few hour or the hole day or at last two so give me that time.

**Comment by Dream-Master** [comment]
to 1. Mushroom generation in TF Flower gen should be on ross or disabled.

**Comment by Chaoschaot232** [comment]
> to 1. Mushroom generation in TF Flower gen should be on ross or disabled.

🤔 Both sounds ok because there are many ways to get the Mushrooms if needed.
But even I havne't used them in any Botania playthrough so I must look them up on my own for what we can use them. If they are not needed mainly like the flowers, disable them for now but let them be craftable. I have an idea for them too ;)

**Comment by Chaoschaot232** [comment]
@Dream-Master From what I have seen Botania Mushrooms are used in exchange for the same colored Petal to craft flowers.
So they are mostly useless and could be disabled entirely.

**Comment by Dream-Master** [comment]
no why. Make them useable instead

**Comment by Ethryan** [comment]
Maybe macerate them into dye dusts instead?

**Comment by Chaoschaot232** [comment]
> Maybe macerate them into dye dusts instead?

This could be a nice use but then I'll reference the GT dusts because why making new ones if we already have them ;)
